### ISSUE TITLE
QuestPDF GeneratePdf method hangs on AWS ECS Fargate

### ISSUE THREAD

**Issue by gwin003** [QuestPDF/QuestPDF#990]
**Describe the bug**
Running QuestPDF on AWS ECS instance hangs when trying to generate PDFs. CPU and memory on the ECS Fargate instance spikes to 100%. Limits are currently set to 4gb of memory and 1 vCPU. 

**To Reproduce**
```
using (var memoryStream = new MemoryStream())
{
  _logger.LogInformation(LogEventIds.GenerateReportProcessing, $"Generating PDF");  //gets logged
  pdfDocument.GeneratePdf(memoryStream);  //hangs here

  _logger.LogInformation(LogEventIds.GenerateReportProcessing, $"MemoryStream.ToArray");  //never gets logged
  fileData = memoryStream.ToArray();
}
```

**Expected behavior**
The memory stream would be set to the `byte[]` representing the report.

**Screenshots**
N/A

**Environment**
```
<PackageReference Include="HarfBuzzSharp.NativeAssets.Linux" Version="7.3.0.2" />
<PackageReference Include="QuestPDF" Version="2024.6.0" />
```
What operating system do you use? (OS type, x64 vs x86 vs arm64)
AWS ECS Fargate - Linux/X86_64

**Additional context**
CPU and Memory spikes, is there a recommended minimum requirements or has someone deployed QuestPDF to AWS/Azure who can share their CPU/Memory usage where its working?

**Comment by MarcinZiabek** [comment]
Thank you for sharing this issue 😄 

Does the problem happen to every generated document? How complex are your documents? Are you using many images or a lot of text instances?

Could you please update to the latest `2024.7.3` release? 

Having any minimal example of the hanging document would be very helpful.

**Comment by v3nus-py** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.

**Comment by aandriichuk** [comment]
Hello. I have the same issue with latest release. Linux based container on Kubernetes but in Azure. Example: 

```
 Document.Create(container =>
            {
                container.Page(page =>
                {
                    page.Size(PageSizes.A4);
                    page.Margin(2, Unit.Centimetre);
                    page.PageColor(Colors.White);
                    page.DefaultTextStyle(x => x.FontSize(20));

                    page.Header()
                        .Text("Hello PDF!")
                        .SemiBold().FontSize(36).FontColor(Colors.Blue.Medium);

                    page.Content()
                        .PaddingVertical(1, Unit.Centimetre)
                        .Column(x =>
                        {
                            x.Spacing(20);

                            x.Item().Text(Placeholders.LoremIpsum());
                            x.Item().Image(Placeholders.Image(200, 100));
                        });

                    page.Footer()
                        .AlignCenter()
                        .Text(x =>
                        {
                            x.Span("Page ");
                            x.CurrentPageNumber();
                        });
                });
            })
            .GeneratePdf();
```

```
<PackageReference Include="QuestPDF" Version="2024.10.4" />
<PackageReference Include="SkiaSharp.NativeAssets.Linux.NoDependencies" Version="2.88.9" />
```

**Comment by aandriichuk** [comment]
I also did the following because of exception from FontManager:

```
QuestPDF.Settings.FontDiscoveryPaths.Clear();
var defaultFontPath = AppDomain.CurrentDomain.BaseDirectory;
QuestPDF.Settings.FontDiscoveryPaths.Add($"{defaultFontPath}Fonts");
```

And I found that library hangs here

`page.Header().Text("Hello PDF!")`

**Comment by MarcinZiabek** [comment]
@aandriichuk, Could you please share what values are returned by the invocations below, on the virtual machine that the library hangs?

```c#
AppDomain.CurrentDomain.RelativeSearchPath
AppDomain.CurrentDomain.BaseDirectory
```

**Comment by aandriichuk** [comment]
@MarcinZiabek 

```
AppDomain.CurrentDomain.RelativeSearchPath = null
AppDomain.CurrentDomain.BaseDirectory = /
```

**Comment by MarcinZiabek** [comment]
Interesting!

Could you please also try the `AppContext.BaseDirectory` value?

In the vast majority of execution environments, all three options should return exactly the same value.

**Comment by aandriichuk** [comment]
@MarcinZiabek 

 `AppContext.BaseDirectory = /`

I am not sure but as far as I understood it seems something wrong with paths and loading default fonts. I am trying to load them manually. Locally `AppDomain.CurrentDomain.RelativeSearchPath` also is null and two others has value `/app/ ` but this value is valid for local env and locally everything works good.

**Comment by MarcinZiabek** [comment]
Indeed, all methods should return the path where the application is stored. However, in certain environments, an incorrect path is being returned for some reason. When the path is null or `/`, the QuestPDF library ends up scanning the entire file system for font files, which, as expected, takes a significant amount of time and causes the application to hang. This behavior appears to be an environment-specific issue rather than a bug in the library. I am exploring ways to retrieve a more reliable path.

Would you like to also give it a try the following list?

```c#
Environment.CurrentDirectory,
Directory.GetCurrentDirectory(),
Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)
```

My proposal is to use `AppContext.BaseDirectory` as the default. If it is null or shorter than 10 characters (or any arbitrary threshold), we could fall back to an alternative method, such as the one mentioned above.

**Comment by aandriichuk** [comment]
@MarcinZiabek all including `AppDomain.CurrentDomain.BaseDirectory` returns /

I would propose to make the path as a required setting and throw exception if it is not specified. Default value could be /AppContext.BaseDirectory (or fallback value)/LatoFont. This should guranty that default fonts always loaded even if not installed in an os. I would avoid the scan of entire bin folder because of when application hangs it is puzzling and you do not have idea what is wrong because of no exceptions and no logs.

**Comment by aandriichuk** [comment]
@MarcinZiabek I've changed my code and it hangs now at another place :) Here is example

       ```
        QuestPDF.Settings.License = LicenseType.Community;
        QuestPDF.Settings.EnableDebugging = true;
        QuestPDF.Settings.FontDiscoveryPaths.Clear();

        var fontsPath = new FileInfo(Assembly.GetExecutingAssembly().Location).Directory!
            .GetDirectories("Fonts")
            .FirstOrDefault();

        var files = fontsPath!.GetFiles("*.ttf");
        foreach (var fileInfo in files)
        {
            using var stream = File.OpenRead(fileInfo.FullName);
            FontManager.RegisterFont(stream);
        }
```
Problem is here `FontManager.RegisterFont(stream);` Is this related to path's issue as well?

**Comment by aandriichuk** [comment]
We've changed a way of how we are deploying and pdf reports are created now :)

**Comment by MarcinZiabek** [comment]
Within the 2024.12.0 release, I have further adjusted the code to reduce the probability of this type of issue. I hope it will offer a better experience for other developers.

Thank you for your help and patience!

**Comment by aandriichuk** [comment]
> Within the 2024.12.0 release, I have further adjusted the code to reduce the probability of this type of issue. I hope it will offer a better experience for other developers.
> 
> Thank you for your help and patience!

Thank you for the great library!

**Comment by MarcinZiabek** [comment]
Since there has been no recent activity, I’m closing this issue. If you have a similar request in the future, feel free to open a new GitHub issue. Thanks for your feedback! ❤️
